dbms/src/Common/UniThreadPool.cpp

```diff
@@ -350,7 +350,7 @@ void ThreadPoolImpl<Thread>::worker(typename std::list<Thread>::iterator thread_
             std::lock_guard lock(mutex);
             --scheduled_jobs;
 
-            if (threads.size() > scheduled_jobs + max_free_threads)
+            if (!shutdown && threads.size() > scheduled_jobs + max_free_threads)
             {
                 thread_it->detach();
```

Comment:
```suggestion
            // If shutdown is called, let the `finalize` clear all threads
            // Otherwise reduce the allocated threads number according to the running jobs and `max_free_threads`
            if (!shutdown && threads.size() > scheduled_jobs + max_free_threads)
            {
                thread_it->detach();
```